### ISSUE TITLE
feature: Convert `Unpack[TypedDict]` to regular arguments

### ISSUE THREAD

**Issue by llucax** [mkdocstrings/python#207]
### Is your feature request related to a problem? Please describe.

Not a problem, but when writing functions/classes with lots of options, and in particular when these options need to be forwarded down to a few levels, it is extremely convenient to use `Unpack[TypeDict]` as the type for keyword arguments and group all arguments in a class that can be reused.

For example:

```py
from typing import TypedDict, Unpack

class Args(TypedDict):
    """Arguments for the add function"""

    a: int
    """The first number"""
    b: int
    """The second number"""

def add(**args: Unpack[Args]) -> int:
    """Add two numbers together.

    Args:
        **args: The arguments for the function.

    Returns:
        The sum of the two numbers.
    """
    return args["a"] + args["b"]
```

Renders as:

![image](https://github.com/user-attachments/assets/07ce28ff-11ed-457f-8d34-c1afa0c0bbbf)

### Describe the solution you'd like

It would be nice if it were rendered as regular arguments instead, like:

![image](https://github.com/user-attachments/assets/2c785d99-8dd6-4815-82bf-828939858fea)

### Describe alternatives you've considered

They are basically above.

### Additional context

I know this might be tricky, specially considering cases using `Args(TypeDict, total=False)` or `NotRequired` as in this case some arguments could not be present at all, which is different from passing `None` for example. But maybe there is a low hanging fruit, and at least some limited support could be added.

**Comment by pawamoy** [comment]
Thanks for the request @llucax! Related: https://github.com/mkdocstrings/griffe/issues/284.

**Comment by llucax** [comment]
Oh, thanks for the reference, I completely missed it. I was actually in doubt if it had to be reported in griffe or not :shrug:

**Comment by pawamoy** [comment]
No worries, it's not a duplicate. Support for `Unpack` must be implemented in Griffe, but here we also want an option to actually unpack the signature when rendering :slightly_smiling_face:

**Comment by Kludex** [comment]
This would be very nice. I was going to ask about it.

**Comment by samuelcolvin** [comment]
This would be great plase @pawamoy.

**Comment by llucax** [comment]
🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥

**Comment by pawamoy** [comment]
Yeah I'll update the feature and go with option 5, adding logic for `total=True/False`, `typing.Required` and `typing.NotRequired`. Non-required parameters will be swallowed by `**kwargs` and documented in an "Other parameters" section.

**Comment by llucax** [comment]
I was also thinking that maybe the best way to document non-required fields with no default with some sort of sentinel default. From the docs PoV that looks like enough, but I'm not sure how this would work in griffe, as I guess it is doing some AST transformations or something like that.

> What do IDEs do?

Here is an example of what pyright does (in vim)

<img width="397" height="321" alt="Image" src="https://github.com/user-attachments/assets/5e64cdb9-64e9-44b0-83e2-3301556496e7" />

<img width="407" height="363" alt="Image" src="https://github.com/user-attachments/assets/346a7a43-3a2d-4941-9f34-01abe2c23853" />

So it uses `...` as a *sentinel* for non required, which is not correct but I find reasonable. It also doesn't rewrite the docstring, but I think mkdocstrings should.

>  I think the most correct thing would be to expand all parameters, but add a default value like MISSING or NOT_REQUIRED to non-required ones.

This seems to be what pyright does.

> We'd only add back `**kwargs` if `extra_items` is specified. But then, what's the description of kwargs? Setting an english-hardcoded description seems ugly (unless we make upstream tools like mkdocstrings translate it, but still ugly). Also we must find a non-colliding name for kwargs (which could be taken by the typed dict already).

What is `extra_items`?

> What do you want to show to your users? What do your users want to see instead of **kwargs: Unpack[Thing]?

I find what pyright does reasonable. The only problem would be if the user accepts `...` as a value that is different from the value being missing, and/or if they use `...` as default for required fields, that would introduce an ambiguity (is this `...` because it is an explicit default or is it because the field is not required). I think in practice `...` should be semantically as missing, so using it for something else would be unexpected anyways, and this can happen with many other type hints, where type-checking can pass but at runtime you could do whatever you want and break all guarantees.

**Comment by pawamoy** [comment]
I went ahead with expanding to only named arguments, for both required and non-required fields, while non-required parameters get a `...` default value. That's instead of adding `**kwargs` for non-required parameters.

The `extra_items` and `closed` features of typed dicts are not yet supported, we'll add support later when people request it (these are features for Python 3.15).
